src/Resources/skeleton/authenticator/Security51LoginFormAuthenticator.tpl.php

```diff
@@ -0,0 +1,74 @@
+<?= "<?php\n" ?>
+
+namespace <?= $namespace ?>;
+
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractLoginFormAuthenticator;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class <?= $class_name; ?> extends AbstractLoginFormAuthenticator
+{
+    use TargetPathTrait;
+
+    public const LOGIN_ROUTE = 'app_login';
+
+    private <?= $use_typed_properties ? 'UrlGeneratorInterface ' : null ?>$urlGenerator;
+
+    public function __construct(UrlGeneratorInterface $urlGenerator)
+    {
+        $this->userRepository = $userRepository;
+        $this->urlGenerator = $urlGenerator;
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return self::LOGIN_ROUTE === $request->attributes->get('_route')
+            && $request->isMethod('POST');
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        /** @var string $password */
+        $password = $request->request->get('password', '');
+
+        /** @var string $userIdentifier */
+        $userIdentifier = $request->request->get('<?= $username_field ?>', '');
+
+        /** @var string|null */
+        $csrf = $request->get('_csrf_token');
+
+        return new Passport(
+            new UserBadge($userIdentifier),
+            new PasswordCredentials($password),
+            [
+                new PasswordUpgradeBadge($password, $this->userRepository),
```

Comment:
```suggestion
                new PasswordUpgradeBadge($password),
```
- Placeholder - 
@wouterj has a PR in the works to eliminate the need to pass a `PasswordUpgraderInterface` as a require 2nd arg.

Reply:
https://github.com/symfony/symfony/pull/39111

Reply:
For reference, the upstream PR was just merged.